### ISSUE TITLE
Slow shape extraction

### ISSUE THREAD

**Issue by francescofarina** [ml-explore/mlx#693]
The speed of extracting the shape of an array seems significantly lower (2-4x) than in other libraries (numpy/pytorch). 

As far as I've seen, performance is independent wrt number of rows and cols and using gpu or cpu doesn't affect the results.

**To reproduce**
```python
import timeit

import mlx.core as mx
import numpy as np
import torch

rows = 10000
cols = 100
a = np.ones([rows, cols])
times = timeit.Timer(lambda: a.shape[0]).repeat(repeat=100)
print("numpy", min(times))
a = torch.ones([rows, cols])
times = timeit.Timer(lambda: a.shape[0]).repeat(repeat=100)
print("pytorch", min(times))
a = mx.ones([rows, cols])
times = timeit.Timer(lambda: a.shape[0]).repeat(repeat=100)
print("mlx", min(times))
```

Results on M3 Pro
```
numpy 0.036397125106304884
pytorch 0.07941891602240503
mlx 0.12359304097481072
```

**Desktop**
- macOS 14.2
- branch: main

**Comment by awni** [comment]
Oh interesting!! Thanks for the benchmark. Probably it's because we are doing a copy of the shape. I think that it should be an easy fix which has been my intention since we changed shape to be a tuple.

**Comment by awni** [comment]
Hmm, this one might not be such an easy fix after all. It seems more related to pybind11 overhead than making a copy.

We were exploring reducing pybind11 overhead using alternatives but it's a lot of work and in most cases very far from being a bottleneck.

I'm curious, did you notice this as a bottleneck in you code?

**Comment by francescofarina** [comment]
Thanks for the update @awni.

To answer your question, no, I don't think it's a major bottleneck. But let me give you a bit more context so that you can evaluate better.

With @TristanBilot we're working on mlx-graphs and doing some benchmarks and profiling to understand where we can improve performance. Now, when dealing with graphs, batching them usually involves collating multiple graphs into a bigger (disconnected) one. While profiling a training loop [here](https://github.com/mlx-graphs/mlx-graphs/blob/917c6b5ec5980554ffd769e5e3bccfaf34569eba/benchmarks/profiling/profile_training.py), as expected we found the collate op as being a relatively expensive one (see below). 

Basically, to perform the collate op, we need to compute the number of nodes in each graph - through profiling we identified that one as a part where we spend an unexpected amount of time (and hence did the analysis for this issue).
Again, it's not a major bottleneck as it account for about 1% of the entire training time, but still it's a bit unexpected and 4x slower than numpy.

<img width="1776" alt="Screenshot 2024-02-16 at 09 38 48" src="https://github.com/ml-explore/mlx/assets/19568231/e5849f96-dac9-4d49-91b5-07bfdc891a8a">

**Comment by TristanBilot** [comment]
The thing is that we call ´.shape[0]´ in a for loop, which generates an overhead. As a workaround, we can use caching so that we compute only once the shape extraction.

**Comment by awni** [comment]
Thanks for the detailed update, that's super useful!

My current best estimate at the moment is that this is caused by pybind11 overhead. I basically replaced the call to `a.shape[0]` in your benchmark with a trivial no-op which goes through pybind and it is about the same speed and a lot slower than NumPy and PyTorch, neither of which use pybind11 (which is known to be slow for bindings).

We investigated switching to nanobind a few weeks ago to get rid of this overhead. It's definitely possible and long-term something we want to do but a big change so hasn't reached the top of the priority stack yet.

**Comment by awni** [comment]
We updated to nanobind #839 and shape extraction is a lot faster (>2x), so I think we can close this. We are still slightly slower than NumPy but I think it is fast enough to not worry about for now. In the posted benchmark I got:


```
numpy 0.03884141705930233
pytorch 0.09683946147561073
mlx 0.058855582028627396
```

**Comment by TristanBilot** [comment]
Good news! Thanks for the update @awni
